File: app/routers/bills.py
"""
SMART BIN LTD - BILLS ROUTER
"""
import uuid
from datetime import date
from decimal import Decimal
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from app.database import get_db
from app.models import Bill, House, Phase, Estate, Resident
from app.schemas import BillCreate, BillBulkCreate, BillOut, BillSummary, Pagination
from app.dependencies import get_current_user, require_role
from app.services.sms_service import sms_service

router = APIRouter(prefix="/api/bills", tags=["bills"])

@router.get("/")
def list_bills(
    status: str = None,
    estate_id: str = None,
    overdue_only: bool = False,
    page: int = 1,
    limit: int = 50,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    offset = (page - 1) * limit

    q = db.query(
        Bill, House.account_number, House.house_number, Phase.phase_number,
        Estate.name.label("estate_name"), Resident.full_name.label("resident_name"),
        Resident.phone_number
    ).join(House, Bill.house_id == House.id)\
     .join(Phase, House.phase_id == Phase.id)\
     .join(Estate, Phase.estate_id == Estate.id)\
     .outerjoin(Resident, (Resident.house_id == House.id) & (Resident.is_primary == True))\
     .filter(Estate.company_id == user.company_id)

    if status:
        q = q.filter(Bill.status == status)
    if estate_id:
        q = q.filter(Estate.id == estate_id)
    if overdue_only:
        q = q.filter(Bill.due_date < func.curdate(), Bill.status != "paid")

    total = q.count()
    rows = q.order_by(Bill.created_at.desc()).offset(offset).limit(limit).all()

    bills = []
    for b, acc, hnum, pnum, ename, rname, rphone in rows:
        d = BillOut.model_validate(b)
        d.account_number = acc
        d.house_number = hnum
        d.phase_number = pnum
        d.estate_name = ename
        d.resident_name = rname
        d.phone_number = rphone
        bills.append(d)

    return {
        "bills": bills,
        "pagination": Pagination(
            page=page, limit=limit, total=total,
            total_pages=(total + limit - 1) // limit
        )
    }

@router.get("/house/{house_id}", response_model=list[BillOut])
def bills_by_house(house_id: str, db: Session = Depends(get_db), user=Depends(get_current_user)):
    return db.query(Bill).filter(Bill.house_id == house_id).order_by(Bill.created_at.desc()).all()

@router.post("/", response_model=BillOut)
def create_bill(req: BillCreate, db: Session = Depends(get_db),
                user=Depends(require_role("super_admin", "admin", "clerk"))):
    bill = Bill(
        id=str(uuid.uuid4()),
        house_id=req.house_id,
        amount=req.amount,
        balance=req.amount,
        bill_period=req.bill_period,
        due_date=req.due_date,
        description=req.description,
        created_by=user.id
    )
    db.add(bill)
    db.commit()
    db.refresh(bill)
    return bill

@router.post("/bulk")
def bulk_create(req: BillBulkCreate, db: Session = Depends(get_db),
                user=Depends(require_role("super_admin", "admin"))):
    q = db.query(House).join(Phase, House.phase_id == Phase.id)\
        .join(Estate, Phase.estate_id == Estate.id)\
        .filter(Estate.company_id == user.company_id, House.status == "active")

    if req.estate_id:
        q = q.filter(Estate.id == req.estate_id)
    if req.phase_id:
        q = q.filter(Phase.id == req.phase_id)

    houses = q.all()
    bills = []
    for h in houses:
        b = Bill(
            id=str(uuid.uuid4()),
            house_id=h.id,
            amount=req.amount,
            balance=req.amount,
            bill_period=req.bill_period,
            due_date=req.due_date,
            description=req.description,
            created_by=user.id
        )
        db.add(b)
        bills.append(b)

    db.commit()
    return {
        "message": f"{len(bills)} bills created successfully",
        "bills_created": len(bills)
    }

@router.get("/summary/overdue", response_model=BillSummary)
def overdue_summary(db: Session = Depends(get_db), user=Depends(get_current_user)):
    result = db.query(
        func.count(Bill.id).label("total_overdue"),
        func.sum(Bill.balance).label("total_amount"),
        func.count(Bill.house_id.distinct()).label("affected_houses")
    ).join(House, Bill.house_id == House.id)\
     .join(Phase, House.phase_id == Phase.id)\
     .join(Estate, Phase.estate_id == Estate.id)\
     .filter(
         Estate.company_id == user.company_id,
         Bill.due_date < func.curdate(),
         Bill.status.in_(["pending", "partial"])
     ).first()

    return BillSummary(
        total_overdue=result.total_overdue or 0,
        total_amount=result.total_amount or Decimal("0.00"),
        affected_houses=result.affected_houses or 0
    )

@router.post("/{bill_id}/send-sms")
def send_bill_sms(bill_id: str, db: Session = Depends(get_db),
                  user=Depends(require_role("super_admin", "admin", "clerk"))):
    bill = db.query(Bill).filter(Bill.id == bill_id).first()
    if not bill:
        raise HTTPException(status_code=404, detail="Bill not found")

    house = db.query(House).filter(House.id == bill.house_id).first()
    resident = db.query(Resident).filter(
        Resident.house_id == house.id,
        Resident.is_primary == True,
        Resident.is_active == True
    ).first()

    if not resident:
        raise HTTPException(status_code=404, detail="No primary resident found for this house")

    estate = db.query(Estate).join(Phase, Phase.estate_id == Estate.id)\
        .join(House, House.phase_id == Phase.id)\
        .filter(House.id == house.id).first()

    result = sms_service.send_bill_sms(
        phone=resident.phone_number,
        name=resident.full_name,
        account_number=house.account_number,
        amount=float(bill.balance),
        period=bill.bill_period,
        due_date=bill.due_date.strftime("%d %b %Y"),
        estate_name=estate.name if estate else "",
        house_id=house.id,
        db=db
    )

    return {"message": "SMS sent", "result": result}

File: src/routes/bills.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const db = require('../config/database');
const { authenticate, authorize } = require('../middleware/auth');
const company = require('../config/company');
const router = express.Router();

// Get all bills (with filters)
router.get('/', authenticate, async (req, res) => {
  try {
    const { status, estate_id, overdue_only, page = 1, limit = 50 } = req.query;
    const offset = (page - 1) * limit;

    let query = `
      SELECT b.*, h.account_number, h.house_number, p.phase_number,
             e.name as estate_name, r.full_name as resident_name, r.phone_number
      FROM bills b
      JOIN houses h ON b.house_id = h.id
      JOIN phases p ON h.phase_id = p.id
      JOIN estates e ON p.estate_id = e.id
      LEFT JOIN residents r ON r.house_id = h.id AND r.is_primary = true
      WHERE e.company_id = $1
    `;
    const params = [req.user.company_id];
    let paramIndex = 2;

    if (status) {
      query += ` AND b.status = $${paramIndex}`;
      params.push(status);
      paramIndex++;
    }

    if (estate_id) {
      query += ` AND e.id = $${paramIndex}`;
      params.push(estate_id);
      paramIndex++;
    }

    if (overdue_only === 'true') {
      query += ` AND b.due_date < CURRENT_DATE AND b.status != 'paid'`;
    }

    query += ` ORDER BY b.created_at DESC LIMIT $${paramIndex} OFFSET $${paramIndex + 1}`;
    params.push(limit, offset);

    const result = await db.query(query, params);

    // Get total count
    const countResult = await db.query(
      `SELECT COUNT(*) FROM bills b
       JOIN houses h ON b.house_id = h.id
       JOIN phases p ON h.phase_id = p.id
       JOIN estates e ON p.estate_id = e.id
       WHERE e.company_id = $1`,
      [req.user.company_id]
    );

    res.json({ 
      bills: result.rows,
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: parseInt(countResult.rows[0].count),
        totalPages: Math.ceil(countResult.rows[0].count / limit)
      }
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Get bills for a specific house
router.get('/house/:houseId', authenticate, async (req, res) => {
  try {
    const result = await db.query(
      `SELECT b.* FROM bills b
       JOIN houses h ON b.house_id = h.id
       WHERE h.id = $1
       ORDER BY b.created_at DESC`,
      [req.params.houseId]
    );
    res.json({ bills: result.rows });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Create single bill
router.post('/', authenticate, authorize('super_admin', 'admin', 'clerk'), async (req, res) => {
  try {
    const { house_id, amount, bill_period, due_date, description } = req.body;
    const id = uuidv4();

    const result = await db.query(
      `INSERT INTO bills (id, house_id, amount, balance, bill_period, due_date, description, created_by) 
       VALUES ($1, $2, $3, $3, $4, $5, $6, $7) RETURNING *`,
      [id, house_id, amount, bill_period, due_date, description, req.user.id]
    );

    res.status(201).json({ bill: result.rows[0] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Bulk create bills for estate/phase
router.post('/bulk', authenticate, authorize('super_admin', 'admin'), async (req, res) => {
  try {
    const { estate_id, phase_id, amount, bill_period, due_date, description } = req.body;

    let housesQuery = `
      SELECT h.id FROM houses h
      JOIN phases p ON h.phase_id = p.id
      JOIN estates e ON p.estate_id = e.id
      WHERE e.company_id = $1 AND h.status = 'active'
    `;
    const params = [req.user.company_id];

    if (estate_id) {
      housesQuery += ` AND e.id = $2`;
      params.push(estate_id);
    }
    if (phase_id) {
      housesQuery += ` AND p.id = $${params.length + 1}`;
      params.push(phase_id);
    }

    const housesResult = await db.query(housesQuery, params);
    const billsCreated = [];

    for (const house of housesResult.rows) {
      const id = uuidv4();
      const result = await db.query(
        `INSERT INTO bills (id, house_id, amount, balance, bill_period, due_date, description, created_by) 
         VALUES ($1, $2, $3, $3, $4, $5, $6, $7) RETURNING *`,
        [id, house.id, amount, bill_period, due_date, description, req.user.id]
      );
      billsCreated.push(result.rows[0]);
    }

    res.status(201).json({ 
      message: `${billsCreated.length} bills created successfully`,
      bills: billsCreated 
    });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

// Get overdue bills summary
router.get('/summary/overdue', authenticate, async (req, res) => {
  try {
    const result = await db.query(
      `SELECT 
        COUNT(*) as total_overdue,
        SUM(balance) as total_amount,
        COUNT(DISTINCT house_id) as affected_houses
       FROM bills b
       JOIN houses h ON b.house_id = h.id
       JOIN phases p ON h.phase_id = p.id
       JOIN estates e ON p.estate_id = e.id
       WHERE e.company_id = $1 AND b.due_date < CURRENT_DATE AND b.status != 'paid'`,
      [req.user.company_id]
    );

    res.json({ summary: result.rows[0] });
  } catch (error) {
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;
